### PR TITLE
MES-995: Refactor state tree to nest captured data inside a particular test (MES-1921)

### PR DESCRIPTION
**Pull request by johnmorleybjss**
## Description and relevant Jira numbers
* Previously, data captured during the test was at the top level of the state (e.g. `/preTestDeclarations`)
* Introduce a top level key, `/tests`
* Under `/tests`, have `currentTest` and `startedTests` keys
* Store each individual test under `startedTests`, key them by journal `slotId` 
* Introduce a new `TestModule`, remove `CandidateModule` and `PreTestDeclarationsModule`
* Create a new `testReducer`, which wraps all test state manipulation actions and is responsible for navigating the correct test's state, then calling reducers for further manipulation

* Depends on https://github.com/dvsa/mes-api-definitions/pull/6/files for schema changes ✅
* This is the alternative approach mentioned on https://github.com/dvsa/mes-mobile-app/pull/214

image

## Pull Request checklist:

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist:

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop
- [x] commits are squashed

## Sign off process checklist:

- [x] Code has been tested manually
- [x] Tests are passing
- [x] PR link added to JIRA
- [x] Reviewers selected in Github
- [ ] Any new reusable component/widget added to component library on Confluence

- [ ] Screenshot(s) captured on the physical device (if applicable)
- [ ] Tested by QA
- [ ] PO's approval

### CODE REVIEW

**Comment by ammarhaiderbjss** [comment]
@johnmorleybjss looks good so far 👍

File: src/modules/test/test.reducer.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      ),
+    },
+    current: {
+      slotId,
+    },
+  };
+};
+
+const deriveSlotId = (state, action): string | null => {
+  if (action instanceof testOutcomeActions.TestOutcomeStartTest) {
```
Comment (ammarhaiderbjss) [inline src/modules/test/test.reducer.ts]: How come we've departed from the standard `switch` statement pattern for reacting to action types??

Comment (johnmorleybjss) [reply src/modules/test/test.reducer.ts]: I think the reason for this departure is that the test reducer is serving a different purpose. The rest of our reducers are "leaf" reducers and are taking actions and directly manipulating one piece of state. This is a parent reducer which delegates most of its manipulation to child reducers. I didn't see much need to `switch`, as all actions are passed to every listed reducer and _the children_ decide whether they care about that action by switching themselves. Perhaps there's a way we could refactor to make it more consistent with our other reducers?

Comment (ammarhaiderbjss) [reply src/modules/test/test.reducer.ts]: Makes perfect sense, this is a common concept in redux. Have a look at this page https://redux.js.org/recipes/structuring-reducers/splitting-reducer-logic, there's a built-in function `combineReducers` that does what you're looking for I think, i.e. splitting up reducers and having them operate on slices of nested state, looks to be available in `ngrx` also - https://ngrx.io/api/store/combineReducers

Comment (johnmorleybjss) [reply src/modules/test/test.reducer.ts]: Managed to simplify the code by using `combineReducers` for the individual test scoped state, but still need the custom wrapper around that to navigate to the correct test.

File: src/modules/tests/tests.reducer.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+) => {
+  const slotId = deriveSlotId(state, action);
+  if (!slotId) {
+    return state;
+  }
+
+  return {
+    ...state,
+    [slotId]: {
+      ...combineReducers(
```
Comment (ammarhaiderbjss) [inline src/modules/tests/tests.reducer.ts]: YES! 😍

File: src/modules/tests/tests.reducer.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import * as testOutcomeActions from '../../pages/journal/components/test-outcome/test-outcome.actions';
+import { preTestDeclarationsReducer } from './pre-test-declarations/pre-test-declarations.reducer';
+import { candidateReducer } from './candidate/candidate.reducer';
+import { get } from 'lodash';
+import { combineReducers } from '@ngrx/store';
+
+// Extend this with any new test domain action types
+type TestAction = preTestDeclarationActions.Types | testOutcomeActions.Types;
+
+const initialState = {};
```
Comment (ammarhaiderbjss) [inline src/modules/tests/tests.reducer.ts]: Can we add some type safety to this?

Comment (johnmorleybjss) [reply src/modules/tests/tests.reducer.ts]: Ah yes, I knew there was another thing I wanted to do! 😁

File: src/modules/tests/tests.reducer.ts (hunk 2 above)
Comment (ammarhaiderbjss) [reply src/modules/tests/tests.reducer.ts]: Can we add a special comment here to explain this esoteric usage of `combineReducers`?

**Comment by szabolcs-keresztes** [comment]
Great solution :)

File: src/modules/tests/tests.reducer.ts (hunk 2 above)
Comment (johnmorleybjss) [reply src/modules/tests/tests.reducer.ts]: Comment added

File: src/modules/tests/tests.reducer.ts (hunk 3 above)
Comment (johnmorleybjss) [reply src/modules/tests/tests.reducer.ts]: Type safety added.

**Pull request merged**